### DISCUSSION TITLE
Auto Approve Mode

### DISCUSSION THREAD

**Discussion by entrepeneur4lyf** [cline/cline#368] [Feature Requests]
Sometimes I wish that I didn’t have to click approve when we are iterating a todo. I frequently have the model make a markdown file with a list of features or bugs or whatever and break them down into small tasks so that we can iterate and stay focused. Cursor will flow through each group of subtasks fairly easily but it’s definitely not as good. I’m not sure how they are handling file structure but the tree sitter seems to help the model though it will hallucinate files. I would like to be able to let it auto-approve for the next x iterations EXCEPT the tool really need to self meter requests per minute. It would be helpful to manage that at any rate.

**Comment by objectiveSee** [comment]
I would love this feature as well! Even if it's just enabled per conversation instead of as a globals setting.

Add a checkbox to the save dialog that says "Auto-save for this thread".

**Comment by theviktor** [comment]
This would only work well if the auto detect truncated code feature has an option to automatically reply to the agent "do not truncate/omit code, respond with full file contents" when detected, since if we have auto approve it would auto approve those faulty truncated edits and mess up the code.

**Reply by objectiveSee** [reply]
How often does that happen? For me I have not seen it happen.

What we could have is when it shows you the save button dialogue, it waits three or four seconds in a countdown and then auto saves

**Comment by jonit-dev** [comment]
Agreed, but for safety reasons, there should be an optional "max time spent on task" input. Otherwise, cline can get stuck in a loop of mistakes, like sometimes it does

**Comment by volome** [comment]
Agree ,but as an option

**Comment by lloydchang** [comment]
Related https://github.com/cline/cline/discussions/368 https://github.com/cline/cline/discussions/635 https://github.com/cline/cline/pull/674 https://github.com/cline/cline/pull/784 https://github.com/cline/cline/pull/802

**Comment by hanzenshou** [comment]
Interested in this as well, would like to see maybe a way to review all deltas at the end and confirm the changes.

**Comment by lloydchang** [comment]
if anyone needs Auto Approve Mode, it's available via Roo Cline at https://marketplace.visualstudio.com/items?itemName=RooVeterinaryInc.roo-cline

**Comment by Gringoloko122** [comment]
Traduza pra mim

**Comment by domdomegg** [comment]
Related: https://github.com/cline/cline/discussions/991

It'd be great to scope this auto-approval to the workspace (rather than allowing Cline to edit any random files on a computer, including say system configuration files / files in other projects without approval).

**Reply by jakenuts** [reply]
If I'm not mistaken the "should I ask for approval" step is a question posed to the model so conceivably you could use the .clinerules workspace file to override the rules either to allow more or less autonomy. I like a fully independent coder with access to anything I would want myself so I add that to my prompt/rules and it seems to work. I'm one timer trigger away from a 24/7 coding assistant.

**Reply by domdomegg** [reply]
This still relies on the model not malfunctioning in making an assessment - I think it would provide more confidence to ensure Cline itself is limiting the permissions boundary (which I believe is the behaviour when auto-approval is not enabled).

**Comment by jakenuts** [comment]
In case you've tuned out - it's here🎉! Cline 3.x now has auto approval with fine grained controls over what can be done autonomously. Also, as the need for approval is determined by the rules as interpreted by the model you can provide overrides in your prompt or .clinerules.

**Comment by saoudrizwan** [comment]
Cline's most requested feature is here 🎉 You can now let Cline work more autonomously with the new Auto-approve menu, letting you choose what tools he can use without needing your permission.

![image](https://github.com/user-attachments/assets/c2cc2e06-abd6-4835-8bdf-0b05283ed7c0)

- Enable or disable for: reading files & directories, editing files, executing terminal commands, using the browser, and using MCP servers.
- Set a limit for how many API requests Cline makes before asking for your approval to keep going.
- And if you have Cline working in the background, he can send system notifications for when he needs your attention, i.e. to get your approval, answer a question, or when a task is completed.

![image](https://github.com/user-attachments/assets/2195f04a-ae37-4dea-a81d-cf9dac5f92a6)
